### DISCUSSION TITLE
How do I schedule jobs transactionally with Exposed’s transaction block?

### DISCUSSION THREAD

**Discussion by Jicol95** [jobrunr/jobrunr#1170] [Q&A] [answered]
Can someone please outline an example of how to schedule jobs only when a transaction commits? I'm using exposed transaction blocks which [looks to be supported with jobrunr pro](https://www.jobrunr.io/en/documentation/pro/transactions/).

Here's a rough example that resulted in the job being scheduled instead of being rolled back.

``` kotlin
fun test(scheduler: JobRequestScheduler) {
  transaction {
    scheduler.schedule(id, Instant.now(), TestJobRequest(id))
    throw SomeException()
  }
}

``` 

I'm aware that I haven't provided much info, but I also don't really know where to start.

**Comment by rdehuyss** [comment]
Hi @Jicol95 , 

We have this test that succeeds:

```kotlin
    @Test
    fun testServiceMethodWithTransactionFails() {
        // GIVEN
        try {
            testService.insertDataAndCreateJobTransactionalThatFails("My String")
            Assertions.fail<Any>("Statement above should have thrown an exception")
        } catch (e: Exception) {
            // expected
        }

        // THEN
        assertThat(getAllJobs()).isEmpty()
    }
```

And `TestService` has the following method
```kotlin
    fun insertDataAndCreateJobTransactionalThatFails(data: String): Int {
        transaction {
            val id: Int = testRepository.insertData(data)
            val jobId = jobScheduler.enqueue { doJobWithData(id) }
            throw RuntimeException("Boem: (recordId=$id / jobId=$jobId)")
        }
    }
```

Which version are you on?

**Reply by Jicol95** [reply]
Thanks for the response 🙌 Your example is the ideal state for me but I'm not quite there yet.

I'm using `v.7.3.0`. In my example I'm using `JobRequestScheduler` rather than `JobScheduler`, are they both "transaction aware"?

**Comment by rdehuyss** [comment]
Can you create a small repo that reproduces the issue? We'll look into it next week.

**Reply by Jicol95** [reply]
Sorry for the delay, took me a bit to set that up to remove anything proprietary from my side. Here's an example repo that attempts to align with how we set things up: https://github.com/Jicol95/transactional-job-runr.

In my case I can see that the job is successfully executed even though my transaction threw an exception. Hopefully it's just a setup issue from my side 🤞 

Output when I run the code
```
Scheduling job with param 2acae106-50b0-4432-9d04-b9b8969c4d0a
SEEEEEYMOOOUUUUUR! 
The job with param 2acae106-50b0-4432-9d04-b9b8969c4d0a did in fact run
```

**Reply by auloin** [reply]
Hi @Jicol95, I don't see any _jobrunr-pro-kotlin-XX-support_  dependency (in your case `jobrunr-pro-kotlin-2.0-support`) listed in your setup. The dependency is required in order for JobRunr Pro to be aware of Exposed transactions.

**Reply by Jicol95** [reply]
~~Ahhh amazing nice catch. Can you please link to that dependency, I cant find it here https://repo.jobrunr.io/ ?~~

~~I only see;~~
~~- jobrunr-pro-kotlin-1.8-support~~
~~- jobrunr-pro-kotlin-1.9-support~~

Found it!

**Reply by Jicol95** [reply]
Adding `jobrunr-pro-kotlin-2.0-support` gives me a new error

```
jicol95/jobrunr-transactions/src/main/kotlin/Main.kt:31:21 Unresolved reference 'saveLicense'.
```

**Reply by rdehuyss** [reply]
Hi @Jicol95 , I wonder where you get the `jobrunr-pro-kotlin-2.0-support` as we've found an error in our `build.gradle` which resulted in it not being published. We've fixed it and added a test that this would not happen again for future kotlin version.

We'll drop you a note as soon as it is released. 

We assume that you've added `jobrunr-kotlin-2.0-support`, bringing in the OSS version of the `StorageProvider` which does not have the `saveLicense` method.

**Reply by Jicol95** [reply]
I took a guess at the URL based on the other releases and it seemed to provide something useable but something is not right for sure in how this is being resolved so I'm doubtful that it is actually working.

`org.jobrunr:jobrunr-kotlin-2.0-support:7.3.0`

so yes your guess was correct, wrong module

**Reply by auloin** [reply, marked as answer]
@Jicol95 we've updated the docs to show how to configure the Exposed transaction plugin: https://www.jobrunr.io/en/documentation/pro/transactions/#configuration.

Please note that, you'll need to wait for the next release to use `jobrunr-pro-kotlin-2.0-support`. A workaround would be to use `jobrunr-pro-kotlin-1.9-support`, which is available right now and also has the transaction plugin for Exposed.

**Reply by Jicol95** [reply]
Incredible stuff 👏 Thanks for all the help and thank you for the documentation!!

**Reply by Jicol95** [reply]
@auloin sorry to revive this but hopefully this is the last piece for me. I've added `org.jobrunr:jobrunr-pro-kotlin-1.9-support:7.3.0` to my project but now I get the following error 

```
> Could not resolve all files for configuration [redacted-internal-project].
   > Could not resolve org.jobrunr:jobrunr-pro-bom:7.3.0.
     Required by:
         project [redacted-internal-project] > [redacted-internal-project] > org.jobrunr:jobrunr-pro-kotlin-1.9-support:7.3.0
```

Looks like the Kotlin support dependency requires the BOM dependency. However when I check https://repo.jobrunr.io/ I don't see any release for `org.jobrunr:jobrunr-pro-bom` with version 7.3.0. Latest is `4.0.3` which blocks me from providing that dependency.

Can you please advise?

**Reply by auloin** [reply]
@Jicol95 could you try with 7.3.2?

**Reply by auloin** [reply]
@Jicol95 you may also try to exclude the `jobrunr-pro-bom` as Gradle doesn't seem to like how the package has been configured. We'll try to fix this for the next release.

```
implementation("org.jobrunr:jobrunr-pro-kotlin-1.9-support:7.3.2") {
    exclude(group = "org.jobrunr", module = "jobrunr-pro-bom")
}
```

**Reply by Jicol95** [reply]
```   
> Could not find org.jobrunr:jobrunr-pro-bom-1.9:7.3.2.
     Searched in the following locations:
          - https://repo.jobrunr.io/private-releases/org/jobrunr/jobrunr-pro-bom-1.9/7.3.2/jobrunr-pro-bom-1.9-7.3.2.pom
```


let me try the other suggestion

**Reply by Jicol95** [reply]
```
implementation("org.jobrunr:jobrunr-pro-kotlin-1.9-support:7.3.2") {
    exclude(group = "org.jobrunr", module = "jobrunr-pro-bom")
}
```


Thanks, this allows my project to build! Any ideas why that's required though and could it be related to there being no release for `jobrunr-pro-bom-1.9-7.3.2`?

**Reply by auloin** [reply]
@Jicol95 tbh, I am not sure exactly why it's required. We use `jobrunr-pro-bom` to internally manage dependency versions, so there is no need for it to be published. This requirement seems to only apply to Gradle managed projects because `jobrunr-pro-kotlin-XX-support` dependencies include a Gradle Module Metadata...

**Reply by auloin** [reply]
Hi @Jicol95, we've released JobRunr Pro 7.4.0. `jobrunr-pro-kotlin-2.0-support` is now available, there is also support for Kotlin 2.1 with `jobrunr-pro-kotlin-2.1-support` and the `jobrunr-pro-bom` is no longer required.

**Comment by Jicol95** [comment]
Something further to add here. Is there a doc outlining what `jobrunr-kotlin-X.X-support` provides?

If not, it would be awesome to see a doc that explains what we unlock by including it. I'd love to share something with my org after I've resolved this to share all of the great work here.
